src/Storages/StorageS3.cpp

```diff
@@ -1264,17 +1270,17 @@ void ReadFromStorageS3Step::initializePipeline(QueryPipelineBuilder & pipeline,
             query_configuration.url.uri.getHost() + std::to_string(query_configuration.url.uri.getPort()),
             iterator_wrapper,
             max_parsing_threads,
-            need_only_count,
-            query_info));
+            need_only_count));
     }
 
-    pipeline.init(Pipe::unitePipes(std::move(pipes)));
-}
+    auto pipe = Pipe::unitePipes(std::move(pipes));
+    if (pipe.empty())
+        pipe = Pipe(std::make_shared<NullSource>(read_from_format_info.source_header));
 
+    for (const auto & processor : pipe.getProcessors())
+        processors.emplace_back(processor);
 
-void ReadFromStorageS3Step::applyFilters()
-{
-    /// We will use filter_dags in filterKeysForPartitionPruning called from initializePipeline, nothing to do here
+    pipeline.init(std::move(pipe));
```

Comment:
Just curious why we need it

Reply:
Please clarify: need what?
The refactoring is needed for analyzer mainly, but it also makes virtual columns analysis in a proper way (using AST is not valid; you can't just take WHERE part cause e.g. it is applied after JOIN).

`applyFilters` itself is needed to build sets for IN which may be needed; and we have to do it before building pipeline

Reply:
I am about change:
```
pipeline.init(Pipe::unitePipes(std::move(pipes)));
```
to
```
auto pipe = Pipe::unitePipes(std::move(pipes));
for (const auto & processor : pipe.getProcessors())
    processors.emplace_back(processor);
pipeline.init(std::move(pipe));
```
Sorry for confusion

Reply:
Ah, this one is for proper EXPLAIN PIPELINE
`processors` is a list of processors we create in the source.